File: MCP/tool-gateway/src/tool-index.js
import { describeConnectionIssue } from "mcporter";

const DEFAULT_REFRESH_INTERVAL_MS = 30_000;
const DEFAULT_RESULT_LIMIT = 5;

/**
 * Maintains an in-memory index of all known MCP tools using mcporter as a backend.
 */
export class ToolIndex {
  constructor(runtime, options = {}) {
    this.runtime = runtime;
    this.refreshIntervalMs = options.refreshIntervalMs ?? DEFAULT_REFRESH_INTERVAL_MS;
    this.logger = options.logger ?? console;
    this.entries = [];
    this.entryMap = new Map();
    this.excludedServers = new Set(options.excludeServers ?? []);
    this.lastRefresh = 0;
    this.refreshPromise = null;
  }

  async ensureFresh(force = false) {
    const needsRefresh =
      force ||
      !this.entries.length ||
      Date.now() - this.lastRefresh > this.refreshIntervalMs;

    if (!needsRefresh) {
      return;
    }

    if (!this.refreshPromise) {
      this.refreshPromise = this.pullLatest()
        .catch((error) => {
          this.logger.error("[tool-index] refresh failed", error);
          throw error;
        })
        .finally(() => {
          this.refreshPromise = null;
        });
    }

    return this.refreshPromise;
  }

  async pullLatest() {
    const servers = this.runtime.listServers();
    const snapshot = [];

    for (const serverName of servers) {
      if (this.excludedServers.has(serverName)) {
        continue;
      }
      try {
        const tools = await this.runtime.listTools(serverName, {
          includeSchema: true,
          autoAuthorize: true,
        });

        for (const tool of tools) {
          snapshot.push(buildEntry(serverName, tool));
        }
      } catch (error) {
        const issue = describeConnectionIssue(error);
        this.logger.warn(
          `[tool-index] Unable to list tools for ${serverName}: ${issue.message ?? error.message}`
        );
      }
    }

    this.entries = snapshot;
    this.entryMap = new Map(snapshot.map((entry) => [entry.id, entry]));
    this.lastRefresh = Date.now();
    return snapshot;
  }

  async search(query, options = {}) {
    const normalizedQuery = normalizeQuery(query);
    if (!normalizedQuery) {
      throw new Error("Query text is required");
    }

    await this.ensureFresh(options.forceRefresh ?? false);

    const tokens = normalizedQuery.split(/\s+/).filter(Boolean);
    const limit = clampCount(options.limit ?? DEFAULT_RESULT_LIMIT);
    const serverHint = options.serverHint?.toLowerCase?.() ?? null;

    const scored = this.entries
      .filter((entry) =>
        serverHint ? entry.serverNormalized.includes(serverHint) : true
      )
      .map((entry) => ({
        entry,
        score: computeScore(entry, tokens, normalizedQuery),
      }))
      .filter((item) => item.score > 0)
      .sort((a, b) => b.score - a.score)
      .slice(0, limit)
      .map(({ entry, score }) => formatMatch(entry, score));

    return {
      matches: scored,
      refreshedAt: new Date(this.lastRefresh).toISOString(),
      totalIndexedTools: this.entries.length,
    };
  }

  async describe(toolId, options = {}) {
    const entry = await this.getTool(toolId, options);
    return formatDescription(entry);
  }

  async getTool(toolId, options = {}) {
    const normalizedId = normalizeToolId(toolId);
    if (!normalizedId) {
      throw new Error("tool_id is required");
    }

    await this.ensureFresh(options.forceRefresh ?? false);
    let entry = this.entryMap.get(normalizedId);

    if (!entry && options.forceRefresh !== true) {
      await this.ensureFresh(true);
      entry = this.entryMap.get(normalizedId);
    }

    if (!entry) {
      throw new Error(`Unknown tool_id "${toolId}"`);
    }

    return entry;
  }

  stats() {
    return {
      totalIndexedTools: this.entries.length,
      refreshedAt: this.lastRefresh ? new Date(this.lastRefresh).toISOString() : null,
      refreshIntervalMs: this.refreshIntervalMs,
    };
  }
}

function buildEntry(serverName, toolInfo) {
  const normalizedDescription = (toolInfo.description ?? "").toLowerCase();
  const normalizedServer = serverName.toLowerCase();
  const normalizedName = toolInfo.name.toLowerCase();
  const inputSummary = summarizeSchema(toolInfo.inputSchema);
  const outputSummary = summarizeSchema(toolInfo.outputSchema);
  const schemaTokens = buildSchemaTokens(inputSummary, outputSummary);
  const risk = classifyRisk(toolInfo.name, toolInfo.description);

  return {
    id: `${serverName}:${toolInfo.name}`,
    server: serverName,
    tool: toolInfo.name,
    description: toolInfo.description ?? null,
    inputSchema: toolInfo.inputSchema ?? null,
    outputSchema: toolInfo.outputSchema ?? null,
    serverNormalized: normalizedServer,
    toolNormalized: normalizedName,
    descriptionNormalized: normalizedDescription,
    schemaTokens,
    inputSummary,
    outputSummary,
    risk,
  };
}

function summarizeSchema(schema) {
  if (!schema || typeof schema !== "object") {
    return null;
  }

  const required = new Set(Array.isArray(schema.required) ? schema.required : []);
  const properties = schema.properties && typeof schema.properties === "object" ? schema.properties : {};
  const fields = Object.entries(properties).map(([name, value]) => ({
    name,
    type: inferSchemaType(value),
    required: required.has(name),
    description:
      value && typeof value === "object" && typeof value.description === "string"
        ? value.description
        : undefined,
  }));

  return {
    type: typeof schema.type === "string" ? schema.type : "object",
    fields,
    required: [...required],
  };
}

function inferSchemaType(node) {
  if (!node || typeof node !== "object") {
    return "unknown";
  }

  if (typeof node.type === "string") {
    return node.type;
  }

  if (Array.isArray(node.type)) {
    return node.type.join("|");
  }

  if (Array.isArray(node.enum)) {
    return `enum(${node.enum.length})`;
  }

  if (node.properties) {
    return "object";
  }

  if (node.items) {
    return "array";
  }

  return "unknown";
}

function buildSchemaTokens(...summaries) {
  return summaries
    .filter(Boolean)
    .flatMap((summary) => summary.fields.map((field) => `${field.name.toLowerCase()} ${field.type}`));
}

const DESTRUCTIVE_KEYWORDS = [
  "delete",
  "destroy",
  "terminate",
  "drop",
  "wipe",
  "revoke",
  "shutdown",
  "disable",
  "remove",
  "kill",
];

const WRITE_KEYWORDS = [
  "create",
  "update",
  "write",
  "send",
  "post",
  "patch",
  "put",
  "start",
  "open",
  "modify",
];

function classifyRisk(name, description) {
  const haystack = `${name ?? ""} ${description ?? ""}`.toLowerCase();
  if (matchesKeyword(haystack, DESTRUCTIVE_KEYWORDS)) {
    return "destructive";
  }
  if (matchesKeyword(haystack, WRITE_KEYWORDS)) {
    return "write";
  }
  return "read";
}

function matchesKeyword(text, keywords) {
  return keywords.some((keyword) => text.includes(keyword));
}

function computeScore(entry, tokens, fallbackQuery) {
  if (!tokens.length) {
    return baseScore(entry, fallbackQuery);
  }

  return tokens.reduce((acc, token) => {
    return (
      acc +
      weightMatch(entry.toolNormalized, token, 4) +
      weightMatch(entry.descriptionNormalized, token, 2) +
      weightMatch(entry.serverNormalized, token, 1) +
      weightSchema(entry.schemaTokens, token)
    );
  }, 0);
}

function baseScore(entry, fallbackQuery) {
  const summaryBlob = [entry.toolNormalized, entry.descriptionNormalized, entry.serverNormalized].join(" ");
  return similarity(summaryBlob, fallbackQuery) * 2;
}

function weightMatch(haystack, needle, multiplier) {
  if (!haystack || !needle) {
    return 0;
  }

  const idx = haystack.indexOf(needle);
  if (idx === -1) {
    return 0;
  }

  const lengthBonus = needle.length / (haystack.length + 1);
  const positionBonus = 1 - idx / (haystack.length + 1);
  return multiplier * (1 + lengthBonus + positionBonus);
}

function weightSchema(tokens, needle) {
  if (!tokens?.length || !needle) {
    return 0;
  }

  let score = 0;
  for (const token of tokens) {
    if (token.includes(needle)) {
      score += 0.5 + needle.length / (token.length + 1);
    }
  }
  return score;
}

function similarity(a, b) {
  if (!a || !b) {
    return 0;
  }

  const shorter = a.length < b.length ? a : b;
  const longer = a.length >= b.length ? a : b;
  let matches = 0;

  for (const char of shorter) {
    if (longer.includes(char)) {
      matches += 1;
    }
  }

  return matches / longer.length;
}

function normalizeQuery(value) {
  return typeof value === "string" ? value.trim().toLowerCase() : "";
}

function clampCount(value) {
  if (!Number.isFinite(value)) {
    return DEFAULT_RESULT_LIMIT;
  }

  return Math.max(1, Math.min(10, Math.floor(value)));
}

function formatMatch(entry, score) {
  return {
    id: entry.id,
    server: entry.server,
    tool: entry.tool,
    description: entry.description,
    score: Number(score.toFixed(3)),
    risk: entry.risk,
    parameters: summarizeParameters(entry.inputSummary),
  };
}

function formatDescription(entry) {
  return {
    id: entry.id,
    server: entry.server,
    tool: entry.tool,
    description: entry.description,
    inputSchema: entry.inputSchema,
    outputSchema: entry.outputSchema,
    inputSummary: entry.inputSummary,
    outputSummary: entry.outputSummary,
    risk: entry.risk,
  };
}

function summarizeParameters(summary) {
  if (!summary?.fields?.length) {
    return [];
  }

  return summary.fields.slice(0, 5).map((field) => ({
    name: field.name,
    type: field.type,
    required: field.required,
  }));
}

function normalizeToolId(value) {
  if (typeof value !== "string") {
    return "";
  }
  return value.trim();
}

File: AIDZeroCode/MCP/tool-gateway/tool_index.py
"""In-memory MCP tool index with lightweight ranking and risk classification."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


DEFAULT_REFRESH_INTERVAL_MS = 30_000
DEFAULT_RESULT_LIMIT = 5
DISCOVERY_QUERIES = {
    "*",
    "all",
    "all tools",
    "available tools",
    "discover tools",
    "list",
    "list available tools",
    "list tools",
    "show tools",
    "tools",
}

DESTRUCTIVE_KEYWORDS = {
    "delete",
    "destroy",
    "terminate",
    "drop",
    "wipe",
    "revoke",
    "shutdown",
    "disable",
    "remove",
    "kill",
}

WRITE_KEYWORDS = {
    "create",
    "update",
    "write",
    "send",
    "post",
    "patch",
    "put",
    "start",
    "open",
    "modify",
}


@dataclass(slots=True)
class ToolEntry:
    id: str
    server: str
    tool: str
    description: str | None
    input_schema: dict[str, Any] | None
    output_schema: dict[str, Any] | None
    server_normalized: str
    tool_normalized: str
    description_normalized: str
    schema_tokens: list[str]
    input_summary: dict[str, Any] | None
    output_summary: dict[str, Any] | None
    risk: str


class ToolIndex:
    def __init__(
        self,
        runtime,
        *,
        refresh_interval_ms: int = DEFAULT_REFRESH_INTERVAL_MS,
        exclude_servers: list[str] | None = None,
        server_list_timeout_seconds: float = 8.0,
    ) -> None:
        self.runtime = runtime
        self.refresh_interval_ms = max(1, int(refresh_interval_ms))
        self._exclude_servers = set(exclude_servers or [])
        self.server_list_timeout_seconds = max(0.5, float(server_list_timeout_seconds))
        self.entries: list[ToolEntry] = []
        self.entry_map: dict[str, ToolEntry] = {}
        self.last_refresh_ms: int = 0

    async def ensure_fresh(self, force: bool = False) -> None:
        now_ms = _now_ms()
        needs_refresh = force or not self.entries or (now_ms - self.last_refresh_ms > self.refresh_interval_ms)
        if needs_refresh:
            await self.pull_latest()

    async def pull_latest(self) -> list[ToolEntry]:
        snapshot: list[ToolEntry] = []
        for server_name in self.runtime.list_servers():
            if server_name in self._exclude_servers:
                continue
            try:
                tools = await self.runtime.list_tools(
                    server_name,
                    include_schema=True,
                    timeout_seconds=self.server_list_timeout_seconds,
                )
            except Exception:
                continue
            for tool in tools:
                snapshot.append(_build_entry(server_name, tool))

        self.entries = snapshot
        self.entry_map = {entry.id: entry for entry in snapshot}
        self.last_refresh_ms = _now_ms()
        return snapshot

    async def search(
        self,
        query: str,
        *,
        limit: int | None = None,
        server_hint: str | None = None,
        force_refresh: bool = False,
    ) -> dict[str, Any]:
        normalized_query = _normalize_query(query)
        if not normalized_query:
            raise ValueError("Query text is required")

        await self.ensure_fresh(force=force_refresh)

        tokens = [token for token in normalized_query.split() if token]
        discovery_mode = _is_discovery_query(normalized_query)
        capped_limit = _clamp_count(limit if limit is not None else DEFAULT_RESULT_LIMIT)
        normalized_server_hint = server_hint.lower() if isinstance(server_hint, str) else None

        scored: list[tuple[ToolEntry, float]] = []
        for entry in self.entries:
            if normalized_server_hint and normalized_server_hint not in entry.server_normalized:
                continue
            if discovery_mode:
                scored.append((entry, 1.0))
                continue
            score = _compute_score(entry, tokens, normalized_query)
            if score > 0:
                scored.append((entry, score))

        if discovery_mode:
            scored.sort(key=lambda item: (item[0].server_normalized, item[0].tool_normalized))
        else:
            scored.sort(key=lambda item: item[1], reverse=True)
        matches = [_format_match(entry, score) for entry, score in scored[:capped_limit]]

        return {
            "matches": matches,
            "refreshedAt": _iso_from_ms(self.last_refresh_ms),
            "totalIndexedTools": len(self.entries),
        }

    async def describe(self, tool_id: str, *, force_refresh: bool = False) -> dict[str, Any]:
        entry = await self.get_tool(tool_id, force_refresh=force_refresh)
        return _format_description(entry)

    async def get_tool(self, tool_id: str, *, force_refresh: bool = False) -> ToolEntry:
        normalized = _normalize_tool_id(tool_id)
        if not normalized:
            raise ValueError("tool_id is required")

        await self.ensure_fresh(force=force_refresh)
        entry = self.entry_map.get(normalized)

        if entry is None and not force_refresh:
            await self.ensure_fresh(force=True)
            entry = self.entry_map.get(normalized)

        if entry is None:
            raise ValueError(f"Unknown tool_id '{tool_id}'")

        return entry

    def stats(self) -> dict[str, Any]:
        return {
            "totalIndexedTools": len(self.entries),
            "refreshedAt": _iso_from_ms(self.last_refresh_ms) if self.last_refresh_ms else None,
            "refreshIntervalMs": self.refresh_interval_ms,
        }


def _build_entry(server_name: str, tool_info: dict[str, Any]) -> ToolEntry:
    description = tool_info.get("description") if isinstance(tool_info.get("description"), str) else None
    tool_name = str(tool_info.get("name", "")).strip()
    input_schema = tool_info.get("inputSchema") if isinstance(tool_info.get("inputSchema"), dict) else None
    output_schema = tool_info.get("outputSchema") if isinstance(tool_info.get("outputSchema"), dict) else None

    input_summary = _summarize_schema(input_schema)
    output_summary = _summarize_schema(output_schema)

    return ToolEntry(
        id=f"{server_name}:{tool_name}",
        server=server_name,
        tool=tool_name,
        description=description,
        input_schema=input_schema,
        output_schema=output_schema,
        server_normalized=server_name.lower(),
        tool_normalized=tool_name.lower(),
        description_normalized=(description or "").lower(),
        schema_tokens=_build_schema_tokens(input_summary, output_summary),
        input_summary=input_summary,
        output_summary=output_summary,
        risk=_classify_risk(tool_name, description or ""),
    )


def _summarize_schema(schema: dict[str, Any] | None) -> dict[str, Any] | None:
    if not isinstance(schema, dict):
        return None

    required = schema.get("required") if isinstance(schema.get("required"), list) else []
    required_set = {str(item) for item in required if isinstance(item, str)}
    properties = schema.get("properties") if isinstance(schema.get("properties"), dict) else {}

    fields: list[dict[str, Any]] = []
    for name, value in properties.items():
        if not isinstance(name, str):
            continue
        node = value if isinstance(value, dict) else {}
        description = node.get("description") if isinstance(node.get("description"), str) else None
        fields.append(
            {
                "name": name,
                "type": _infer_schema_type(node),
                "required": name in required_set,
                "description": description,
            }
        )

    return {
        "type": schema.get("type") if isinstance(schema.get("type"), str) else "object",
        "fields": fields,
        "required": sorted(required_set),
    }


def _infer_schema_type(node: dict[str, Any]) -> str:
    node_type = node.get("type")
    if isinstance(node_type, str):
        return node_type
    if isinstance(node_type, list):
        values = [str(item) for item in node_type]
        return "|".join(values)
    if isinstance(node.get("enum"), list):
        return f"enum({len(node['enum'])})"
    if isinstance(node.get("properties"), dict):
        return "object"
    if "items" in node:
        return "array"
    return "unknown"


def _build_schema_tokens(*summaries: dict[str, Any] | None) -> list[str]:
    tokens: list[str] = []
    for summary in summaries:
        if not isinstance(summary, dict):
            continue
        fields = summary.get("fields") if isinstance(summary.get("fields"), list) else []
        for field in fields:
            if not isinstance(field, dict):
                continue
            name = str(field.get("name", "")).lower().strip()
            field_type = str(field.get("type", "")).strip()
            if name:
                tokens.append(f"{name} {field_type}")
    return tokens


def _classify_risk(name: str, description: str) -> str:
    haystack = f"{name} {description}".lower()
    if any(keyword in haystack for keyword in DESTRUCTIVE_KEYWORDS):
        return "destructive"
    if any(keyword in haystack for keyword in WRITE_KEYWORDS):
        return "write"
    return "read"


def _compute_score(entry: ToolEntry, tokens: list[str], fallback_query: str) -> float:
    if not tokens:
        return _base_score(entry, fallback_query)

    score = 0.0
    for token in tokens:
        score += _weight_match(entry.tool_normalized, token, 4.0)
        score += _weight_match(entry.description_normalized, token, 2.0)
        score += _weight_match(entry.server_normalized, token, 1.0)
        score += _weight_schema(entry.schema_tokens, token)
    return score


def _base_score(entry: ToolEntry, fallback_query: str) -> float:
    blob = " ".join([entry.tool_normalized, entry.description_normalized, entry.server_normalized])
    return _similarity(blob, fallback_query) * 2.0


def _weight_match(haystack: str, needle: str, multiplier: float) -> float:
    if not haystack or not needle:
        return 0.0
    idx = haystack.find(needle)
    if idx == -1:
        return 0.0
    length_bonus = len(needle) / (len(haystack) + 1)
    position_bonus = 1 - idx / (len(haystack) + 1)
    return multiplier * (1 + length_bonus + position_bonus)


def _weight_schema(tokens: list[str], needle: str) -> float:
    if not tokens or not needle:
        return 0.0
    score = 0.0
    for token in tokens:
        if needle in token:
            score += 0.5 + len(needle) / (len(token) + 1)
    return score


def _similarity(a: str, b: str) -> float:
    if not a or not b:
        return 0.0
    shorter, longer = (a, b) if len(a) < len(b) else (b, a)
    matches = sum(1 for char in shorter if char in longer)
    return matches / max(1, len(longer))


def _normalize_query(value: Any) -> str:
    return value.strip().lower() if isinstance(value, str) else ""


def _is_discovery_query(query: str) -> bool:
    normalized = " ".join(query.split())
    return normalized in DISCOVERY_QUERIES


def _clamp_count(value: Any) -> int:
    try:
        numeric = int(value)
    except (TypeError, ValueError):
        return DEFAULT_RESULT_LIMIT
    return max(1, min(10, numeric))


def _format_match(entry: ToolEntry, score: float) -> dict[str, Any]:
    return {
        "id": entry.id,
        "server": entry.server,
        "tool": entry.tool,
        "description": entry.description,
        "score": round(score, 3),
        "risk": entry.risk,
        "parameters": _summarize_parameters(entry.input_summary),
    }


def _format_description(entry: ToolEntry) -> dict[str, Any]:
    return {
        "id": entry.id,
        "server": entry.server,
        "tool": entry.tool,
        "description": entry.description,
        "inputSchema": entry.input_schema,
        "outputSchema": entry.output_schema,
        "inputSummary": entry.input_summary,
        "outputSummary": entry.output_summary,
        "risk": entry.risk,
    }


def _summarize_parameters(summary: dict[str, Any] | None) -> list[dict[str, Any]]:
    if not isinstance(summary, dict):
        return []
    fields = summary.get("fields") if isinstance(summary.get("fields"), list) else []
    out: list[dict[str, Any]] = []
    for field in fields[:5]:
        if not isinstance(field, dict):
            continue
        out.append(
            {
                "name": str(field.get("name", "")),
                "type": str(field.get("type", "unknown")),
                "required": bool(field.get("required", False)),
            }
        )
    return out


def _normalize_tool_id(value: Any) -> str:
    return value.strip() if isinstance(value, str) else ""


def _now_ms() -> int:
    return int(datetime.now(tz=timezone.utc).timestamp() * 1000)


def _iso_from_ms(value: int) -> str:
    return datetime.fromtimestamp(value / 1000, tz=timezone.utc).isoformat()
